### ISSUE TITLE
Thanks for the interesting paper! Will the source code of the student network and the learning pipeline of this network be posted? 

### ISSUE THREAD

**Issue by diwert-ai** [weiqingq/CLRKDNet#3]
?

**Comment by diwert-ai** [comment]
I kind of figured it out. As I understood the essence of neural network architecture in neck/agg.py . I wrote a training pipelines with loss function as in the article and it seems to really work! You can close the issue. Thanks! Very informative method to improve the performance of CLR models.

**Comment by XenonJack** [comment]
> I kind of figured it out. As I understood the essence of neural network architecture in neck/agg.py . I wrote a training pipelines with loss function as in the article and it seems to really work! You can close the issue. Thanks! Very informative method to improve the performance of CLR models.

@diwert-ai Hi, I'm also interested in how the training pipeline works, would you like to share your codes if possible? that would be very helpful

**Comment by diwert-ai** [comment]
Hi! The logic goes something like that:
```python

from tqdm import tqdm
import torch
import torch.nn.functional as F
from mmcv.parallel import MMDataParallel

from clrnet.utils.config import Config as TeacherConfig
from clrnet.datasets.process.process import Process as TeacherProcess
from clrnet.models.registry import build_net as teacher_build_net
from clrnet.datasets import build_dataloader as teacher_build_dataloader

from clrkd.utils.config import Config as StudentConfig
from clrkd.datasets.process.process import Process as StudentProcess
from clrkd.models.registry import build_net as student_build_net
from clrkd.engine.optimizer import build_optimizer as student_build_optimizer
from clrkd.engine.scheduler import build_scheduler as student_build_scheduler


def to_cuda(batch):
    for k in batch:
        if not isinstance(batch[k], torch.Tensor):
            continue
        batch[k] = batch[k].cuda()
    return batch

def calculate_attention_map(x):
    # Compute the attention map as the normalized mean of the squared outputs across the channel dimension
    attention_map = x.pow(2).mean(dim=1, keepdim=True)
    normalized_attention_map = F.normalize(attention_map, p=2, dim=[2, 3])
    return normalized_attention_map


if __name__ == '__main__':
    teacher_cfg_path = ...
    student_cfg_path = ...
    model_path = ...

    teacher_cfg = TeacherConfig.fromfile(teacher_cfg_path)
    teacher_processes = TeacherProcess(teacher_cfg.val_process, teacher_cfg)
    teacher_net = teacher_build_net(teacher_cfg)
    teacher_net = MMDataParallel(teacher_net, device_ids=range(1)).cuda()
    teacher_net.load_state_dict(torch.load(model_path)['net'], strict=False)
    teacher_net.eval()

    student_cfg = StudentConfig.fromfile(student_cfg_path)
    batch_size = student_cfg.batch_size
    student_processes = StudentProcess(student_cfg.val_process, student_cfg)
    student_net = student_build_net(student_cfg)
    student_net = MMDataParallel(student_net, device_ids=range(1)).cuda()
    student_optimizer = student_build_optimizer(student_cfg, student_net)
    student_scheduler = student_build_scheduler(student_cfg, student_optimizer)
    train_loader = teacher_build_dataloader(teacher_cfg.dataset.train,
                                            teacher_cfg,
                                            is_train=True)
    
    # train loop
    for epoch in range(student_cfg.epochs):
        pbar = tqdm(train_loader)

        # train epoch loop
        student_net.train()
        for step, data in enumerate(pbar):
            input = to_cuda(data)
            teacher_att_maps = []

            with torch.no_grad():
                teacher_backbone_acts = teacher_net.module.backbone(input['img'])
                teacher_logits = teacher_net(input)
                for act in teacher_backbone_acts:
                    att_map = calculate_attention_map(act)
                    teacher_att_maps.append(calculate_attention_map(act))
                teacher_priors = teacher_net.module.heads.prior_embeddings.weight

            _, student_att_maps = student_net.module.backbone(input['img'], att_flag=True)
            student_logits, clr_out = student_net(input)
            clr_loss = clr_out['loss'].sum()
            student_priors = student_net.module.heads.prior_embeddings.weight
            att_loss = torch.Tensor([0.0]).cuda()
            for i in range(len(teacher_att_maps)):
                att_loss += (teacher_att_maps[i]-student_att_maps[i]).norm()
            prior_loss = (student_priors - teacher_priors).norm(dim=1).mean()
            # logit_loss = ((student_logits - teacher_logits)**2).mean()
            loss = (# 5 * logit_loss + 
                    3 * prior_loss + 
                    att_loss + 
                    clr_loss)
            loss.backward()

            student_optimizer.step()
            student_scheduler.step()
            student_optimizer.zero_grad()

        

        # validation loop
        student_net.eval()

        # visual validation and store model
        pass
        
        # metric validation and logging (tensor board?)   
        pass 

```

In my case (specific dataset) logit distillation made the training unstable, so that part of the loss was commented out.

**Comment by weiqingq** [comment]
The official training code is now available. Thank you, @diwert-ai, for your valuable insights on applying the model to different datasets.
